### PR TITLE
Use Knuth sampler for rate sampling

### PR DESCRIPTION
**Pull request by randomanderson**
This pull request changes the default mechanism of sampling from "random" to "Knuth's algorithm".  This deterministic algorithm is used by the go, Ruby, dotnet, Python, and c++ tracers.

Additionally, I have deprecated some old sampling code in `AbstractSampler` for later removal

### CODE REVIEW

File: dd-trace-ot/src/test/groovy/datadog/trace/api/sampling/KnuthSamplerTest.groovy [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+import datadog.opentracing.DDSpan
+import datadog.trace.common.sampling.KnuthSampler
+import datadog.trace.util.test.DDSpecification
+
+class KnuthSamplerTest extends DDSpecification {
+
+  def "test known values: #traceId"() {
+    given:
+    KnuthSampler sampler = new KnuthSampler(0.5)
```
Comment (tylerbenson) [inline dd-trace-ot/src/test/groovy/datadog/trace/api/sampling/KnuthSamplerTest.groovy]: Can we see some test cases with rate of `0` and `1`?

File: dd-trace-ot/src/main/java/datadog/trace/common/sampling/RandomSampler.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package datadog.trace.common.sampling;
+
+import datadog.opentracing.DDSpan;
+import java.util.concurrent.ThreadLocalRandom;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class RandomSampler implements RateSampler {
```
Comment (tylerbenson) [inline dd-trace-ot/src/main/java/datadog/trace/common/sampling/RandomSampler.java]: Is this actually being used?

File: dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+import datadog.opentracing.DDSpan;
+import datadog.opentracing.DDTracer;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class KnuthSampler implements RateSampler {
+  private static final BigInteger HASH_CONSTANT = new BigInteger("1111111111111111111");
```
Comment (tylerbenson) [inline dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java]: I don't understand the purpose of this.

File: dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  @Override
+  public boolean sample(final DDSpan span) {
+    final boolean sampled;
+    if (rate == 1) {
+      sampled = true;
+    } else if (rate == 0) {
+      sampled = false;
+    } else {
+      sampled = span.getTraceId().multiply(HASH_CONSTANT).mod(MODULUS).compareTo(cutoff) < 0;
```
Comment (tylerbenson) [inline dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java]: Please add a comment explaining the math here.

File: dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package datadog.trace.common.sampling;
+
+import datadog.opentracing.DDSpan;
+import datadog.opentracing.DDTracer;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class KnuthSampler implements RateSampler {
```
Comment (tylerbenson) [inline dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java]: Any nice links you could include that describe the algorithm?

File: dd-trace-ot/src/main/java/datadog/trace/common/sampling/RandomSampler.java (hunk 2 above)
Comment (randomanderson) [reply dd-trace-ot/src/main/java/datadog/trace/common/sampling/RandomSampler.java]: Originally I was going to perhaps have a configuration option to choose but never ended up doing that.  Just adds more complexity.

Removed.

File: dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java (hunk 4 above)
Comment (randomanderson) [reply dd-trace-ot/src/main/java/datadog/trace/common/sampling/KnuthSampler.java]: There's no explanation.  This is a port of the other implementations (that also have no explanation).

See:
https://github.com/DataDog/dd-trace-go/blob/v1/ddtrace/tracer/sampler.go#L85https://github.com/DataDog/dd-trace-py/blob/e4bfdd8d5eb318af9c034452c9e1a1d41d0cdd3b/ddtrace/sampler.py#L59
https://github.com/DataDog/dd-trace-dotnet/blob/e0ed4604a9478a9e995255e6a5a0c145e08ecf36/src/Datadog.Trace/Sampling/RuleBasedSampler.cs#L100

I did rename "HASH_CONSTANT" to "KNUTH_FACTOR" to use the same nomenclature as the other implementations

File: dd-trace-ot/src/test/groovy/datadog/trace/api/sampling/KnuthSamplerTest.groovy (hunk 1 above)
Comment (randomanderson) [reply dd-trace-ot/src/test/groovy/datadog/trace/api/sampling/KnuthSamplerTest.groovy]: Added.

**Comment by randomanderson** [comment]
Creating a new pull request with the other sampling changes.

**Pull request closed without merging**